File: Graphs C++/14 - Strongly Connected Components/topological Order.cpp
#include<bits/stdc++.h>

using namespace std;

const int N = 1e5 + 1;

vector<int> gr[N];
vector<int> order;
int vis[N];

void dfs(int cur, int par) {
	vis[cur] = 1;
	for (auto x : gr[cur]) {
		if (!vis[x]) {
			dfs(x, cur);
		}
	}
	// I am here because I came back from the subtree
	order.push_back(cur);
	return;
}


int main()
{
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);

	int n, m;
	cin >> n >> m;
	for (int i = 0; i < m; i++) {
		int x, y;
		cin >> x >> y;
		gr[x].push_back(y);
	}

	for (int i = 1; i <= n; i++) {
		if (!vis[i]) dfs(i, 0);
	}

	reverse(order.begin(), order.end());

	for (auto x : order) cout << x << " ";

	return 0;
}

File: Graphs Java/src/StronglyConnectedComponent/TopologicalOrder.java
package StronglyConnectedComponent;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class TopologicalOrder {

    static final int N = (int)(1e5 + 1);

    static ArrayList<Integer>[] gr = new ArrayList[N];
    static ArrayList<Integer> order;
    static {
        for(int i = 0 ; i < N ; i++){
            gr[i] = new ArrayList<>();
        }
    }

    static boolean[] vis = new boolean[N];

    static void dfs(int cur, int par) {
        vis[cur] = true;
        for (int x : gr[cur]) {
            if (!vis[x]) {
                dfs(x, cur);
            }
        }
        // I am here because I came back from the subtree
        order.add(cur);
        return;
    }


    public static void main(String[] args) {

        Scanner scn = new Scanner(System.in);
        int n = scn.nextInt(), m = scn.nextInt();

        for (int i = 0; i < m; i++) {
            int x = scn.nextInt(), y = scn.nextInt();
            gr[x].add(y);

        }


        for (int i = 1; i <= n; i++) {
            if (!vis[i]) dfs(i, 0);
        }

        Collections.reverse(order);

        for (int x : order)
            System.out.println(x);

    }
}
